_posts/2018-12-04-imposing-invariants-in-scala.md

```diff
@@ -0,0 +1,359 @@
+---
+layout: post
+author: cmteixeira
+title: Imposing invariants in Scala
+---
+
+
+
+Imagine a opera house allows reservations to a show. For each reservation two people can come: The buyer of the reservation himself, and his guest.  
+As in any show-house, there are many types of seats; some are standing, some normal and some on vip tribunes.
+
+
+
+As a restriction from the opera house, both buyer and guest must be on the same seating area. For example, it is not permitted for one ticket to be standing while the other for the normal seated area.  
+I believe this is what people call an invariant. A requirement that must hold for the data to have meaning.  
+How to enforce then this constraint ?  
+Let us first define the data types relevant to the problem:  
+
+
+```scala
+sealed trait Seat
+
+case class NormalSeat(row: Int, column: Int) extends Seat
+case class VipTribune(tribuneId: String) extends Seat
+case object Standing extends Seat
+```
+
+In what follows it will be useful to discern the type relation above on the following diagram.
+
+<img class=" size-full wp-image-397 aligncenter" src="https://aerodatablog.files.wordpress.com/2018/12/typehiearchy1.png" alt="ActionBuilder" />   
```

Comment:
Don't like the colors.

Reply:
I think the arrows are the wrong way, they should be pointing away from `Seat` as they extend `Seat` and not the other way round.

Reply:
See answer from **artamonovdev** at https://stackoverflow.com/questions/1874049/explanation-of-the-uml-arrows.